### ISSUE TITLE
FP4 inference for 4 bit

### ISSUE THREAD

**Issue by antirez** [antirez/ds4#74]
Right now we have Q4_K GGUF, however FP4 is the way the model is shipped and is numerically different than Q4_K which is uniform quantization. Given that llama.cpp is able to do the inference of GPT120B OSS (I believe) fast enough, it should be possible to support at least as *optional format* the FP4 weights, in order to really do the inference of the Real Thing that DeepSeek shipped. I doubt there are large differences as even the 2 bit quants work well, but... still. Not a priority but something to remember it is worth investigating.

**Comment by ggerganov** [comment]
Btw, I'm running some AIME2025 evals of `ds4` using the 2-bit quant in order to compare against the same eval against the DeepSeek API (i.e. as a reference). I think the run will be ready sometime tomorrow and will share the results (takes ~8 hours on my M2 Ultra). It's not a comprehensive eval by any means, but it might give us some extra information about how well the quality is preserved after the quantization.

**Comment by antirez** [comment]
@ggerganov thank you su much! I uploaded imatrix 2 bit weights too, but I just found that the imatrix for the down side of the routed experts was computed in a wrong way, so... regenerating from scratch. Even the potentially broken version had a lot less drift compared to the 4 bit quants logits, and more exact prefix sharing in generations. But indeed without benchmarks it is hard to tell.

**Comment by ggerganov** [comment]
@antirez The AIME2025x4 eval is still running, but it's already clear that even the 2-bit quant scores close to 100% on that eval. This is a good indication for the quality/correctness of the implementation. The few failing tests in the log are because of the 64k output limit that I used for running the eval (otherwise, it would have to run even longer on some of the problems that require very long thinking).

<details>

```
Processing 120 AIME2025 tasks ...
Servers (1):
  1. mac — http://127.0.0.1:8000 (1 threads)
Model: deepseek-v4-flash
Grader: llm
Sampling: temp=skip, top-k=skip, top-p=skip, min-p=skip

         Task ID              Dataset    Prompt (first 40 chars)                           Expected   Answer     Tokens   Status

  1/120  aime2025_000_020     AIME2025   Circle $\omega_1$ with radius 6 centered at...    293        293        8186     ✓  [  1/  1, 1.000]  mac
  2/120  aime2025_000_006     AIME2025   The twelve letters $A,B,C,D,E,F,G,H,I,J,K$,...    821        821        18845    ✓  [  2/  2, 1.000]  mac
  3/120  aime2025_000_008     AIME2025   The parabola with equation $y=x^{2}-4$ is r...    62         62         19212    ✓  [  3/  3, 1.000]  mac
  4/120  aime2025_000_004     AIME2025   There are $8!=40320$ eight-digit positive i...    279        279        8606     ✓  [  4/  4, 1.000]  mac
  5/120  aime2025_000_015     AIME2025   Six points $ A, B, C, D, E, $ and $ F $ lie...    468        468        1691     ✓  [  5/  5, 1.000]  mac
  6/120  aime2025_000_028     AIME2025   Let $ \triangle ABC $ be a right triangle w...    104        104        16669    ✓  [  6/  6, 1.000]  mac
  7/120  aime2025_000_013     AIME2025   Let $ABCDE$ be a convex pentagon with $AB=1...    60         N/A        65384    ✗  [  6/  7, 0.857]  mac
  8/120  aime2025_000_023     AIME2025   There are $ n $ values of $ x $ in the inte...    149        149        10904    ✓  [  7/  8, 0.875]  mac
  9/120  aime2025_000_022     AIME2025   From an unlimited supply of 1-cent coins, 1...    610        610        20312    ✓  [  8/  9, 0.889]  mac
 10/120  aime2025_000_019     AIME2025   Suppose $ \triangle ABC $ has angles $ \ang...    336        336        30698    ✓  [  9/ 10, 0.900]  mac
 11/120  aime2025_000_012     AIME2025   Alex divides a disk into four quadrants wit...    204        204        37151    ✓  [ 10/ 11, 0.909]  mac
 12/120  aime2025_000_029     AIME2025   There are exactly three positive real numbe...    240        240        22704    ✓  [ 11/ 12, 0.917]  mac
 13/120  aime2025_000_009     AIME2025   The 27 cells of a $3\times9$ grid are fille...    81         81         29234    ✓  [ 12/ 13, 0.923]  mac
 14/120  aime2025_000_010     AIME2025   A piecewise linear periodic function is def...    259        259        11999    ✓  [ 13/ 14, 0.929]  mac
 15/120  aime2025_000_005     AIME2025   An isosceles trapezoid has an inscribed cir...    504        504        2334     ✓  [ 14/ 15, 0.933]  mac
 16/120  aime2025_000_016     AIME2025   Find the sum of all positive integers $ n $...    49         49         1001     ✓  [ 15/ 16, 0.938]  mac
 17/120  aime2025_000_017     AIME2025   Four unit squares form a $2 \times 2$ grid....    82         82         9906     ✓  [ 16/ 17, 0.941]  mac
 18/120  aime2025_000_026     AIME2025   Let $ A_1A_2 \ldots A_{11} $ be an 11-sided...    19         19         9814     ✓  [ 17/ 18, 0.944]  mac
 19/120  aime2025_000_007     AIME2025   Let $k$ be real numbers such that the syste...    77         77         3113     ✓  [ 18/ 19, 0.947]  mac
 20/120  aime2025_000_011     AIME2025   The set of points in 3-dimensional coordina...    510        510        36729    ✓  [ 19/ 20, 0.950]  mac
 21/120  aime2025_000_027     AIME2025   Let the sequence of rationals $ x_1, x_2, \...    248        248        59753    ✓  [ 20/ 21, 0.952]  mac
 22/120  aime2025_000_025     AIME2025   Let $ S $ be the set of vertices of a regul...    113        113        13818    ✓  [ 21/ 22, 0.955]  mac
 23/120  aime2025_000_021     AIME2025   Let $ A $ be the set of positive integer di...    237        237        3683     ✓  [ 22/ 23, 0.957]  mac
 24/120  aime2025_000_001     AIME2025   On $\triangle ABC$ points $A,D,E$, and $B$ ...    588        588        9801     ✓  [ 23/ 24, 0.958]  mac
 25/120  aime2025_000_018     AIME2025   The product $ \prod_{k=4}^{63} \frac{\log_k...    106        106        3510     ✓  [ 24/ 25, 0.960]  mac
 26/120  aime2025_000_002     AIME2025   The 9 members of a baseball team went to an...    16         16         2183     ✓  [ 25/ 26, 0.962]  mac
 27/120  aime2025_000_000     AIME2025   Find the sum of all integer bases $b>9$ for...    70         70         653      ✓  [ 26/ 27, 0.963]  mac
 28/120  aime2025_000_003     AIME2025   Find the number of ordered pairs $(x,y)$, w...    117        117        2127     ✓  [ 27/ 28, 0.964]  mac
 29/120  aime2025_000_014     AIME2025   Let $N$ denote the number of ordered triple...    735        N/A        65438    ✗  [ 27/ 29, 0.931]  mac
 30/120  aime2025_000_024     AIME2025   Sixteen chairs are arranged in a row. Eight...    907        907        10459    ✓  [ 28/ 30, 0.933]  mac
 31/120  aime2025_001_015     AIME2025   Six points $ A, B, C, D, E, $ and $ F $ lie...    468        468        1461     ✓  [ 29/ 31, 0.935]  mac
 32/120  aime2025_001_025     AIME2025   Let $ S $ be the set of vertices of a regul...    113        113        16798    ✓  [ 30/ 32, 0.938]  mac
 33/120  aime2025_001_022     AIME2025   From an unlimited supply of 1-cent coins, 1...    610        610        19975    ✓  [ 31/ 33, 0.939]  mac
 34/120  aime2025_001_006     AIME2025   The twelve letters $A,B,C,D,E,F,G,H,I,J,K$,...    821        821        24063    ✓  [ 32/ 34, 0.941]  mac
 35/120  aime2025_001_000     AIME2025   Find the sum of all integer bases $b>9$ for...    70         70         1441     ✓  [ 33/ 35, 0.943]  mac
 36/120  aime2025_001_011     AIME2025   The set of points in 3-dimensional coordina...    510        510        14654    ✓  [ 34/ 36, 0.944]  mac
 37/120  aime2025_001_024     AIME2025   Sixteen chairs are arranged in a row. Eight...    907        907        14735    ✓  [ 35/ 37, 0.946]  mac
 38/120  aime2025_001_012     AIME2025   Alex divides a disk into four quadrants wit...    204        204        43152    ✓  [ 36/ 38, 0.947]  mac
 39/120  aime2025_001_009     AIME2025   The 27 cells of a $3\times9$ grid are fille...    81         81         34634    ✓  [ 37/ 39, 0.949]  mac
 40/120  aime2025_001_003     AIME2025   Find the number of ordered pairs $(x,y)$, w...    117        117        2204     ✓  [ 38/ 40, 0.950]  mac
 41/120  aime2025_001_013     AIME2025   Let $ABCDE$ be a convex pentagon with $AB=1...    60         N/A        65384    ✗  [ 38/ 41, 0.927]  mac
 42/120  aime2025_001_005     AIME2025   An isosceles trapezoid has an inscribed cir...    504        504        1592     ✓  [ 39/ 42, 0.929]  mac
 43/120  aime2025_001_029     AIME2025   There are exactly three positive real numbe...    240        240        30118    ✓  [ 40/ 43, 0.930]  mac
 44/120  aime2025_001_004     AIME2025   There are $8!=40320$ eight-digit positive i...    279        279        11264    ✓  [ 41/ 44, 0.932]  mac
 45/120  aime2025_001_010     AIME2025   A piecewise linear periodic function is def...    259        259        18172    ✓  [ 42/ 45, 0.933]  mac
 46/120  aime2025_001_020     AIME2025   Circle $\omega_1$ with radius 6 centered at...    293        293        6052     ✓  [ 43/ 46, 0.935]  mac
 47/120  aime2025_001_008     AIME2025   The parabola with equation $y=x^{2}-4$ is r...    62         62         11884    ✓  [ 44/ 47, 0.936]  mac
 48/120  aime2025_001_001     AIME2025   On $\triangle ABC$ points $A,D,E$, and $B$ ...    588        588        5335     ✓  [ 45/ 48, 0.938]  mac
 49/120  aime2025_001_018     AIME2025   The product $ \prod_{k=4}^{63} \frac{\log_k...    106        106        5913     ✓  [ 46/ 49, 0.939]  mac
 50/120  aime2025_001_016     AIME2025   Find the sum of all positive integers $ n $...    49         49         1471     ✓  [ 47/ 50, 0.940]  mac
 51/120  aime2025_001_023     AIME2025   There are $ n $ values of $ x $ in the inte...    149        149        13571    ✓  [ 48/ 51, 0.941]  mac
 52/120  aime2025_001_026     AIME2025   Let $ A_1A_2 \ldots A_{11} $ be an 11-sided...    19         19         7334     ✓  [ 49/ 52, 0.942]  mac
...
```
</details>

**Comment by antirez** [comment]
That's an awesome news @ggerganov, many of us had that feeling playing wit it, but this acknowledge is gold. Btw I'm quite confident that we will see new v4 Flash checkpoints in the future, and potentially even a "coder" version if rumors are true. This can make the model even more useful for the local space. My fingers are crossed. So many thanks for the benchmark: I'm thinking about including at least a few questions in the ds4 test in order to have some real world validation that one could optionally run.

**Comment by txleafeye** [comment]
> I uploaded imatrix 2 bit weights too, but I just found that the imatrix for the down side of the routed experts was computed in a wrong way, so... regenerating from scratch.

Do you plan to update the q2 model or is it only that the uploaded imatrix was computed incorrectly?

**Comment by antirez** [comment]
@txleafeye I uploaded the new quants on HF and the download script was modified as well.

**Comment by txleafeye** [comment]
@antirez Thanks, I misinterpreted the [...]-imatrix.gguf as the imatrix itself, rather than the gguf quantized using the imatrix.

**Comment by antirez** [comment]
I also uploaded the imatrix themselves on HF. The prompts used to generate them are in this repository under gguf-tools.
